File: medium/dp/jump_game.py
"""
Problem 55
"""
from typing import List


class Solution:
    def canJump(self, nums: List[int]) -> bool:

        if len(nums) <= 1:
            return True

        for index in range(1, len(nums) - 1):
            if nums[index - 1] < 1:
                return False
            if nums[index] > nums[index - 1] - 1:
                continue
            else:
                nums[index] = nums[index - 1] - 1

        return nums[len(nums) - 2] > 0

File: src/main/java/leet/solutions/JumpGame.java
package leet.solutions;

/*
https://leetcode.com/problems/jump-game/
Leetcode#55
 */


public class JumpGame {

    public boolean solve(int[] nums)
    {
        Solution solution = new Solution();
        return solution.canJump(nums);
    }

    class Solution {
        public boolean canJump(int[] nums) {

            if (nums.length==1)
            {
                return true;
            }

            for (int i=1;i<nums.length;i++)
            {
                if (nums[i-1] <= 0)
                {
                    return false;
                }
                nums[i] = Math.max(nums[i-1]-1, nums[i]);
            }

            return true;
        }
    }


}
